ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -341,20 +382,30 @@ IoRegisterPlugPlayNotification(IN IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
         case EventCategoryTargetDeviceChange:
         {
             Entry->FileObject = (PFILE_OBJECT)EventCategoryData;
+            Status = IoGetRelatedTargetDevice((PFILE_OBJECT)EventCategoryData, &Entry->FileObjectPDO);
+
+            if (!NT_SUCCESS(Status))
+            {
+                ExFreePoolWithTag(Entry, TAG_PNP_NOTIFY);
+                ObDereferenceObject(DriverObject);
+                return Status;
+            }
             break;
         }
         default:
         {
             DPRINT1("%s: unknown EventCategory 0x%x UNIMPLEMENTED\n",
                     __FUNCTION__, EventCategory);
-            break;
+
+            ExFreePoolWithTag(Entry, TAG_PNP_NOTIFY);
+            ObDereferenceObject(DriverObject);
+            return STATUS_NOT_SUPPORTED;
         }
     }
 
-    KeAcquireGuardedMutex(&PnpNotifyListLock);
-    InsertHeadList(&PnpNotifyListHead,
-                   &Entry->PnpNotifyList);
-    KeReleaseGuardedMutex(&PnpNotifyListLock);
+    ExInterlockedInsertTailList(&PiAddedNotifyListHead,
```

Comment:
I don't really understand the point of having 2 lists. Why do you need to add it to a different list, when you don't hold any locks while processing an entry from the main list? If you want to add a new entry, just add it the tail and it will be processed at the end of the loop that walks the list, instead of at the head as it was done previously.

Reply:
The idea was to make adding and removing operations lock-free.

> just add it the tail and it will be processed at the end of the loop that walks the list

How can it work if the new element is added while we process the last element? How to be with deletions?